README
Description:

5calc is an HTML5-based calculator with built-in theming abilities.


Chrome app:

https://chrome.google.com/webstore/detail/jcbncbadgbbcfnlfmhjbndifohmeckib/


Usage:

5calc has been tested and is working in Firefox 3.6+, Chrome 8+, and Opera 11+, although it will probably work in many other browsers.


Requirements:

No extra requirements as of yet.


Keyboard Shortcuts:

+ ......... Add
- ......... Subtract
/ ......... Divide
* ......... Multiply
x ......... Multiply
= ......... Calculate
Enter ..... Calculate
Escape .... Clear
Backspace . Backspace
Delete .... Backspace
( ......... Left paren
) ......... Right paren
. ......... Decimal
0 ......... Zero
1 ......... One
2 ......... Two
3 ......... Three
4 ......... Four
5 ......... Five
6 ......... Six
7 ......... Seven
8 ......... Eight
9 ......... Nine
s ......... Sine
S ......... Arcsine
c ......... Cosine
C ......... Arccosine
t ......... Tangent
T ......... Arctangent
| ......... Pipe (absolute value)
l ......... Log (base 10)
n ......... Natural log (base E)
! ......... Factorial
r ......... Square root == x^(1/2)
% ......... Modulo (remainder)
^ ......... Exponent
p ......... Pi
e ......... E


CHANGES
=== 0.2.3.2 ===

* Fixed bug with backspace key on delete keypress


=== 0.2.3.1 ===

* Fixed malformed expression with new precision fix


=== 0.2.3 ===

* Fixed precision issues (hopefully)

* Added some additional keyboard shortcuts

* Added space around operators on display


=== 0.2.2 ===

* Fixed bug where multiply and divide don't work correctly in basic mode


=== 0.2.1 ===

* Fixed bug where asin, acos, and atan were not calculated correctly in degrees mode


=== 0.2 ===

* Added advanced mode

* Fixed bug where navigation link colors were not updated with preferences change


=== 0.1 ===

* Initial release to Chrome webstore, basic mode only


js/5calc.js
window.onload = function() {
    apply_settings(get_storage());
    document.onkeypress = key_press;
    document.onkeydown = key_down;
    document.onkeyup = key_up;
};

function key_press(evt, div_id) {
    var id = div_id ? div_id : '';
    var c = evt ? evt.charCode : -1;
    var data;

    if (c == 61 || id == 'equals') {
        button_click('equals', 3);
        calculate();
        scroll_left();
        return true;
    }
    else if (c == 48 || id == 'zero')
        data = ['zero', '0'];
    else if (c == 49 || id == 'one')
        data = ['one', '1'];
    else if (c == 50 || id == 'two')
        data = ['two', '2'];
    else if (c == 51 || id == 'three')
        data = ['three', '3'];
    else if (c == 52 || id == 'four')
        data = ['four', '4'];
    else if (c == 53 || id == 'five')
        data = ['five', '5'];
    else if (c == 54 || id == 'six')
        data = ['six', '6'];
    else if (c == 55 || id == 'seven')
        data = ['seven', '7'];
    else if (c == 56 || id == 'eight')
        data = ['eight', '8'];
    else if (c == 57 || id == 'nine')
        data = ['nine', '9'];
    else if (c == 47 || id == 'divide')
        data = ['divide', '÷'];
    else if (c == 42 || id == 'multiply')
        data = ['multiply', '×'];
    else if (c == 45 || id == 'subtract')
        data = ['subtract', '-'];
    else if (c == 46 || id == 'dot')
        data = ['dot', '.'];
    else if (c == 43 || id == 'add')
        data = ['add', '+'];
    else if (c == 94 || id == 'power')
        data = ['power', '^'];
    else if (c == 40 || id == 'lparen')
        data = ['lparen', '('];
    else if (c == 41 || id == 'rparen')
        data = ['rparen', ')'];
    else
        return true;

    button_click(data[0], 3);
    append_data(data[1]);
    scroll_right();
    return false;
}

function key_down(evt, div_id) {
    var id = div_id ? div_id : '';
    var k = evt ? evt.keyCode : -1;

    if (k == 27 || id == 'clr') { // escape
        button_click('clr', 1);
        if (id)
            setTimeout("button_click('clr', 2)", 150);
        reset();
    }
    else if (k == 8 || id == 'bksp') { // backspace
        button_click('bksp', 1);
        if (id)
            setTimeout("button_click('bksp', 2)", 150);
        cut_data();
    }
    else if (k == 13) { // enter
        button_click('equals', 1);
        calculate();
        scroll_left();
    }
}

function key_up(evt) {
    var k = evt.keyCode;

    if (k == 27) // escape
        button_click('clr', 2);
    else if (k == 8) // backspace
        button_click('bksp', 2);
    else if (k == 13) // enter
        button_click('equals', 2);
}

function calculate() {
    var disp = document.getElementById('display');
    var error = document.getElementById('error');
    var result = __parse(disp.innerHTML);

    if (result.err > 0)
        error.innerHTML = 'Malformed expression';
    else {
        error.innerHTML = '';
        disp.innerHTML = result.res;
    }
}

function reset() {
    document.getElementById('display').innerHTML = '0';
    document.getElementById('error').innerHTML = '';
}

function scroll_right() {
    var disp = document.getElementById('display');
    disp.scrollLeft = disp.scrollWidth;
}

function scroll_left() {
    document.getElementById('display').scrollLeft = 0;
}

function append_data(str) {
    var disp = document.getElementById('display');

    if (disp.innerHTML == '0')
        disp.innerHTML = str;
    else
        disp.innerHTML += str;
}

function cut_data() {
    var disp = document.getElementById('display');

    disp.innerHTML = disp.innerHTML.substring(0, disp.innerHTML.length - 1);
    if (disp.innerHTML == '')
        disp.innerHTML = '0';
}

function button_click(button_id, mode) {
    var btn = document.getElementById(button_id);
    var colors = get_storage();

    switch (mode) {
        case 1: // button down
            btn.style.backgroundColor = '#' + colors.btnfont;
            btn.style.color = '#' + colors.btn;
            break;
        case 2: // button up
            btn.style.backgroundColor = '#' + colors.btn;
            btn.style.color = '#' + colors.btnfont;
            break;
        case 3: // button down, then up
            button_click(button_id, 1);
            setTimeout("button_click('" + button_id + "', 2)", 150);
            break;
    }
}

function cancel_settings() {
    document.getElementById('settings').style.display = 'none';
    apply_settings(get_storage());
}

function save_settings() {
    document.getElementById('settings').style.display = 'none';
    var colors = {
        bg: document.getElementById('bgcolor').value,
        shadow: document.getElementById('shadowcolor').value,
        calc: document.getElementById('calccolor').value,
        disp: document.getElementById('dispcolor').value,
        dispfont: document.getElementById('dispfontcolor').value,
        errorfont: document.getElementById('errorfontcolor').value,
        btn: document.getElementById('btncolor').value,
        btnfont: document.getElementById('btnfontcolor').value,
    };
    set_storage(colors);
    apply_settings(colors);
}

function apply_settings(colors) {
    document.body.style.backgroundColor = '#' + colors.bg;

    if (document.body.style.MozBoxShadow == '') {
        document.getElementById('navbar').style.MozBoxShadow = '0px 0px 10px #' + colors.shadow;
        document.getElementById('inner').style.MozBoxShadow = '0px 0px 10px #' + colors.shadow;
        document.getElementById('settings').style.MozBoxShadow = '0px 0px 10px #' + colors.shadow;
    }
    else if (document.body.style.WebkitBoxShadow == '') {
        document.getElementById('navbar').style.WebkitBoxShadow = '0px 0px 10px #' + colors.shadow;
        document.getElementById('inner').style.WebkitBoxShadow = '0px 0px 10px #' + colors.shadow;
        document.getElementById('settings').style.WebkitBoxShadow = '0px 0px 10px #' + colors.shadow;
    }
    else if (document.body.style.boxShadow == '') {
        document.getElementById('navbar').style.boxShadow = '0px 0px 10px #' + colors.shadow;
        document.getElementById('inner').style.boxShadow = '0px 0px 10px #' + colors.shadow;
        document.getElementById('settings').style.boxShadow = '0px 0px 10px #' + colors.shadow;
    }

    document.getElementById('inner').style.backgroundColor = '#' + colors.calc;
    document.getElementById('display').style.backgroundColor = '#' + colors.disp;
    document.getElementById('display').style.color = '#' + colors.dispfont;
    document.getElementById('error').style.color = '#' + colors.errorfont;

    document.getElementById('navbar').style.backgroundColor = '#' + colors.btn;
    document.getElementById('navbar').style.color = '#' + colors.btnfont;
    document.getElementById('settings').style.backgroundColor = '#' + colors.btn;
    document.getElementById('settings').style.color = '#' + colors.btnfont;
    var btns = document.getElementsByClassName('button');
    for (var x = 0; x < btns.length; x++) {
        btns[x].style.backgroundColor = '#' + colors.btn;
        btns[x].style.color = '#' + colors.btnfont;
    }
}

function show_settings(colors) {
    if (!colors)
        colors = get_storage();

    var bg = document.getElementById('bgcolor');
    var shadow = document.getElementById('shadowcolor');
    var calc = document.getElementById('calccolor');
    var disp = document.getElementById('dispcolor');
    var dispfont = document.getElementById('dispfontcolor');
    var errorfont = document.getElementById('errorfontcolor');
    var btn = document.getElementById('btncolor');
    var btnfont = document.getElementById('btnfontcolor');
    var settings = document.getElementById('settings');

    bg.value = colors.bg;
    bg.style.backgroundColor = '#' + colors.bg;
    bg.style.color = '#' + inverse(colors.bg);

    shadow.value = colors.shadow;
    shadow.style.backgroundColor = '#' + colors.shadow;
    shadow.style.color = '#' + inverse(colors.shadow);

    calc.value = colors.calc;
    calc.style.backgroundColor = '#' + colors.calc;
    calc.style.color = '#' + inverse(colors.calc);

    disp.value = colors.disp;
    disp.style.backgroundColor = '#' + colors.disp;
    disp.style.color = '#' + inverse(colors.disp);

    dispfont.value = colors.dispfont;
    dispfont.style.backgroundColor = '#' + colors.dispfont;
    dispfont.style.color = '#' + inverse(colors.dispfont);

    errorfont.value = colors.errorfont;
    errorfont.style.backgroundColor = '#' + colors.errorfont;
    errorfont.style.color = '#' + inverse(colors.errorfont);

    btn.value = colors.btn;
    btn.style.backgroundColor = '#' + colors.btn;
    btn.style.color = '#' + inverse(colors.btn);

    btnfont.value = colors.btnfont;
    btnfont.style.backgroundColor = '#' + colors.btnfont;
    btnfont.style.color = '#' + inverse(colors.btnfont);

    settings.style.display = 'block';
    settings.style.top = Math.round(
            (window.innerHeight - settings.offsetHeight) / 2) + 'px';
    color_changed();
}

function inverse(c) {
    if (c.length != 6 || ! c.match(/^[0-9a-fA-F]{6}$/))
        return '';

    var r = 255 - parseInt(c.slice(0, 2), 16);
    var g = 255 - parseInt(c.slice(2, 4), 16);
    var b = 255 - parseInt(c.slice(4, 6), 16);

    return r.toString(16) + g.toString(16) + b.toString(16);
}

function set_storage(colors) {
    localStorage.bg = colors.bg;
    localStorage.shadow = colors.shadow;
    localStorage.calc = colors.calc;
    localStorage.disp = colors.disp;
    localStorage.dispfont = colors.dispfont;
    localStorage.errorfont = colors.errorfont;
    localStorage.btn = colors.btn;
    localStorage.btnfont = colors.btnfont;
}

function get_storage() {
    var def = get_defaults();
    return {
        bg: localStorage.bg ? localStorage.bg : def.bg,
        shadow: localStorage.shadow ? localStorage.shadow : def.shadow,
        calc: localStorage.calc ? localStorage.calc : def.calc,
        disp: localStorage.disp ? localStorage.disp : def.disp,
        dispfont: localStorage.dispfont ? localStorage.dispfont : def.dispfont,
        errorfont: localStorage.errorfont ? localStorage.errorfont : def.errorfont,
        btn: localStorage.btn ? localStorage.btn : def.btn,
        btnfont: localStorage.btnfont ? localStorage.btnfont : def.btnfont,
    };
}

function get_defaults() {
    return {
        bg: 'E0E0E0',
        shadow: '000000',
        calc: 'C0C0C0',
        disp: 'FFFFFF',
        dispfont: '000000',
        errorfont: 'FF0000',
        btn: '8C8C8C',
        btnfont: '000000',
    };
}

function highlight_button(btn) {
    var colors = get_storage();
    btn.style.backgroundColor = '#' + colors.bg;
}

function unhighlight_button(btn) {
    var colors = get_storage();
    btn.style.backgroundColor = '#' + colors.btn;
}

function color_changed() {
    var colors = {
        bg: document.getElementById('bgcolor').value,
        shadow: document.getElementById('shadowcolor').value,
        calc: document.getElementById('calccolor').value,
        disp: document.getElementById('dispcolor').value,
        dispfont: document.getElementById('dispfontcolor').value,
        errorfont: document.getElementById('errorfontcolor').value,
        btn: document.getElementById('btncolor').value,
        btnfont: document.getElementById('btnfontcolor').value,
    }
    apply_settings(colors);
}


js/basic.js
function key_press(evt, div_id) {
    var id = div_id ? div_id : '';
    var c = evt ? evt.charCode : -1;
    var data;

    if (c == 61 || id == 'equals') {
        button_click('equals', 3);
        calculate();
        scroll_left();
        return true;
    }
    else if (c == 48 || id == 'zero')
        data = ['zero', '0'];
    else if (c == 49 || id == 'one')
        data = ['one', '1'];
    else if (c == 50 || id == 'two')
        data = ['two', '2'];
    else if (c == 51 || id == 'three')
        data = ['three', '3'];
    else if (c == 52 || id == 'four')
        data = ['four', '4'];
    else if (c == 53 || id == 'five')
        data = ['five', '5'];
    else if (c == 54 || id == 'six')
        data = ['six', '6'];
    else if (c == 55 || id == 'seven')
        data = ['seven', '7'];
    else if (c == 56 || id == 'eight')
        data = ['eight', '8'];
    else if (c == 57 || id == 'nine')
        data = ['nine', '9'];
    else if (c == 47 || id == 'divide')
        data = ['divide', ' / '];
    else if (c == 42 || c == 120 || id == 'multiply')
        data = ['multiply', ' * '];
    else if (c == 45 || id == 'subtract')
        data = ['subtract', ' - '];
    else if (c == 46 || id == 'dot')
        data = ['dot', '.'];
    else if (c == 43 || id == 'add')
        data = ['add', ' + '];
    else if (c == 40 || id == 'lparen')
        data = ['lparen', '('];
    else if (c == 41 || id == 'rparen')
        data = ['rparen', ')'];
    else
        return true;

    button_click(data[0], 3);
    append_data(data[1]);
    scroll_right();
    return false;
}

function key_down(evt, div_id) {
    var id = div_id ? div_id : '';
    var k = evt ? evt.keyCode : -1;

    if (k == 27 || id == 'clr') { // escape
        button_click('clr', 1);
        if (id)
            setTimeout("button_click('clr', 2)", 150);
        reset();
    }
    else if (k == 8 || k == 46 || id == 'bksp') { // backspace/delete
        button_click('bksp', 1);
        if (id)
            setTimeout("button_click('bksp', 2)", 150);
        cut_data();
    }
    else if (k == 13) { // enter
        button_click('equals', 1);
        calculate();
        scroll_left();
    }
}

function key_up(evt) {
    var k = evt.keyCode;

    if (k == 27) // escape
        button_click('clr', 2);
    else if (k == 8 || k == 46) // backspace/delete
        button_click('bksp', 2);
    else if (k == 13) // enter
        button_click('equals', 2);
}


js/common.js
window.onload = function() {
    apply_settings(get_storage());
    document.onkeypress = key_press;
    document.onkeydown = key_down;
    document.onkeyup = key_up;
};

function calculate() {
    var disp = document.getElementById('display');
    var error = document.getElementById('error');
    var colors = get_storage();
    var degrees = colors.mode == 'deg' ? true : false;
    var result = __parse(disp.innerHTML, degrees);

    if (result.err > 0)
        error.innerHTML = 'Malformed expression';
    else {
        error.innerHTML = '';
        var res = result.res.toFixed(10).toString();
        while (res[res.length - 1] == '0') {
            res = res.substring(0, res.length - 1);
        }
        if (res[res.length - 1] == '.')
            res = res.substring(0, res.length - 1);
        disp.innerHTML = res;
    }
}

function factorial(n) {
    if ((n == 0) || (n == 1))
        return 1;
    else
        return (n * factorial(n - 1));
}

function reset() {
    document.getElementById('display').innerHTML = '0';
    document.getElementById('error').innerHTML = '';
}

function scroll_right() {
    var disp = document.getElementById('display');
    disp.scrollLeft = disp.scrollWidth;
}

function scroll_left() {
    document.getElementById('display').scrollLeft = 0;
}

function append_data(str) {
    var disp = document.getElementById('display');

    if (disp.innerHTML == '0') {
        if (str.length == 3 && str[0] == ' ' && str[2] == ' ')
            disp.innerHTML = str[1];
        else
            disp.innerHTML = str;
    }
    else if ((disp.innerHTML[disp.innerHTML.length - 1] == '(' ||
                disp.innerHTML[disp.innerHTML.length - 1] == '|' ) &&
                str.length == 3 && str[0] == ' ' && str[2] == ' ')
        disp.innerHTML += str[1];
    else
        disp.innerHTML += str;
}

function cut_data() {
    var disp = document.getElementById('display');

    if (disp.innerHTML[disp.innerHTML.length - 1] == ' ' &&
            disp.innerHTML[disp.innerHTML.length - 3] == ' ')
        disp.innerHTML = disp.innerHTML.substring(0, disp.innerHTML.length - 3);
    else
        disp.innerHTML = disp.innerHTML.substring(0, disp.innerHTML.length - 1);

    if (disp.innerHTML == '')
        disp.innerHTML = '0';
}

function button_click(button_id, mode) {
    var btn = document.getElementById(button_id);
    var colors = get_storage();

    switch (mode) {
        case 1: // button down
            btn.style.backgroundColor = '#' + colors.btnfont;
            btn.style.color = '#' + colors.btn;
            break;
        case 2: // button up
            btn.style.backgroundColor = '#' + colors.btn;
            btn.style.color = '#' + colors.btnfont;
            break;
        case 3: // button down, then up
            button_click(button_id, 1);
            setTimeout("button_click('" + button_id + "', 2)", 150);
            break;
    }
}

function cancel_settings() {
    document.getElementById('settings').style.display = 'none';
    apply_settings(get_storage());
}

function save_settings() {
    document.getElementById('settings').style.display = 'none';
    var colors = {
        bg: document.getElementById('bgcolor').value,
        shadow: document.getElementById('shadowcolor').value,
        calc: document.getElementById('calccolor').value,
        disp: document.getElementById('dispcolor').value,
        dispfont: document.getElementById('dispfontcolor').value,
        errorfont: document.getElementById('errorfontcolor').value,
        btn: document.getElementById('btncolor').value,
        btnfont: document.getElementById('btnfontcolor').value,
        mode: document.getElementById('mode').value,
    };
    set_storage(colors);
    apply_settings(colors);
}

function apply_settings(colors) {
    document.body.style.backgroundColor = '#' + colors.bg;

    if (document.body.style.MozBoxShadow == '') {
        document.getElementById('navbar').style.MozBoxShadow = '0px 0px 10px #' + colors.shadow;
        document.getElementById('inner').style.MozBoxShadow = '0px 0px 10px #' + colors.shadow;
        document.getElementById('settings').style.MozBoxShadow = '0px 0px 10px #' + colors.shadow;
    }
    else if (document.body.style.WebkitBoxShadow == '') {
        document.getElementById('navbar').style.WebkitBoxShadow = '0px 0px 10px #' + colors.shadow;
        document.getElementById('inner').style.WebkitBoxShadow = '0px 0px 10px #' + colors.shadow;
        document.getElementById('settings').style.WebkitBoxShadow = '0px 0px 10px #' + colors.shadow;
    }
    else if (document.body.style.boxShadow == '') {
        document.getElementById('navbar').style.boxShadow = '0px 0px 10px #' + colors.shadow;
        document.getElementById('inner').style.boxShadow = '0px 0px 10px #' + colors.shadow;
        document.getElementById('settings').style.boxShadow = '0px 0px 10px #' + colors.shadow;
    }

    document.getElementById('inner').style.backgroundColor = '#' + colors.calc;
    document.getElementById('display').style.backgroundColor = '#' + colors.disp;
    document.getElementById('display').style.color = '#' + colors.dispfont;
    document.getElementById('error').style.color = '#' + colors.errorfont;

    document.getElementById('navbar').style.backgroundColor = '#' + colors.btn;
    document.getElementById('navbar').style.color = '#' + colors.btnfont;
    document.getElementById('settings').style.backgroundColor = '#' + colors.btn;
    document.getElementById('settings').style.color = '#' + colors.btnfont;
    var btns = document.getElementsByClassName('button');
    for (var x = 0; x < btns.length; x++) {
        btns[x].style.backgroundColor = '#' + colors.btn;
        btns[x].style.color = '#' + colors.btnfont;
    }
    var links = document.getElementsByTagName('a');
    for (var x = 0; x < links.length; x++) {
        links[x].style.color = '#' + colors.btnfont;
    }
}

function show_settings(colors) {
    if (!colors)
        colors = get_storage();

    var bg = document.getElementById('bgcolor');
    var shadow = document.getElementById('shadowcolor');
    var calc = document.getElementById('calccolor');
    var disp = document.getElementById('dispcolor');
    var dispfont = document.getElementById('dispfontcolor');
    var errorfont = document.getElementById('errorfontcolor');
    var btn = document.getElementById('btncolor');
    var btnfont = document.getElementById('btnfontcolor');
    var settings = document.getElementById('settings');
    var mode = document.getElementById('mode');

    bg.value = colors.bg;
    bg.style.backgroundColor = '#' + colors.bg;
    bg.style.color = '#' + inverse(colors.bg);

    shadow.value = colors.shadow;
    shadow.style.backgroundColor = '#' + colors.shadow;
    shadow.style.color = '#' + inverse(colors.shadow);

    calc.value = colors.calc;
    calc.style.backgroundColor = '#' + colors.calc;
    calc.style.color = '#' + inverse(colors.calc);

    disp.value = colors.disp;
    disp.style.backgroundColor = '#' + colors.disp;
    disp.style.color = '#' + inverse(colors.disp);

    dispfont.value = colors.dispfont;
    dispfont.style.backgroundColor = '#' + colors.dispfont;
    dispfont.style.color = '#' + inverse(colors.dispfont);

    errorfont.value = colors.errorfont;
    errorfont.style.backgroundColor = '#' + colors.errorfont;
    errorfont.style.color = '#' + inverse(colors.errorfont);

    btn.value = colors.btn;
    btn.style.backgroundColor = '#' + colors.btn;
    btn.style.color = '#' + inverse(colors.btn);

    btnfont.value = colors.btnfont;
    btnfont.style.backgroundColor = '#' + colors.btnfont;
    btnfont.style.color = '#' + inverse(colors.btnfont);

    mode.value = colors.mode;

    settings.style.display = 'block';
    settings.style.top = Math.round(
            (window.innerHeight - settings.offsetHeight) / 2) + 'px';
    color_changed();
}

function inverse(c) {
    if (c.length != 6 || ! c.match(/^[0-9a-fA-F]{6}$/))
        return '';

    var r = 255 - parseInt(c.slice(0, 2), 16);
    var g = 255 - parseInt(c.slice(2, 4), 16);
    var b = 255 - parseInt(c.slice(4, 6), 16);

    return r.toString(16) + g.toString(16) + b.toString(16);
}

function set_storage(colors) {
    localStorage.bg = colors.bg;
    localStorage.shadow = colors.shadow;
    localStorage.calc = colors.calc;
    localStorage.disp = colors.disp;
    localStorage.dispfont = colors.dispfont;
    localStorage.errorfont = colors.errorfont;
    localStorage.btn = colors.btn;
    localStorage.btnfont = colors.btnfont;
    localStorage.mode = colors.mode;
}

function get_storage() {
    var def = get_defaults();
    return {
        bg: localStorage.bg ? localStorage.bg : def.bg,
        shadow: localStorage.shadow ? localStorage.shadow : def.shadow,
        calc: localStorage.calc ? localStorage.calc : def.calc,
        disp: localStorage.disp ? localStorage.disp : def.disp,
        dispfont: localStorage.dispfont ? localStorage.dispfont : def.dispfont,
        errorfont: localStorage.errorfont ? localStorage.errorfont : def.errorfont,
        btn: localStorage.btn ? localStorage.btn : def.btn,
        btnfont: localStorage.btnfont ? localStorage.btnfont : def.btnfont,
        mode: localStorage.mode ? localStorage.mode : def.mode,
    };
}

function get_defaults() {
    return {
        bg: 'E0E0E0',
        shadow: '000000',
        calc: 'C0C0C0',
        disp: 'FFFFFF',
        dispfont: '000000',
        errorfont: 'FF0000',
        btn: '8C8C8C',
        btnfont: '000000',
        mode: 'rad',
    };
}

function highlight_button(btn) {
    var colors = get_storage();
    btn.style.backgroundColor = '#' + colors.bg;
}

function unhighlight_button(btn) {
    var colors = get_storage();
    btn.style.backgroundColor = '#' + colors.btn;
}

function color_changed() {
    var colors = {
        bg: document.getElementById('bgcolor').value,
        shadow: document.getElementById('shadowcolor').value,
        calc: document.getElementById('calccolor').value,
        disp: document.getElementById('dispcolor').value,
        dispfont: document.getElementById('dispfontcolor').value,
        errorfont: document.getElementById('errorfontcolor').value,
        btn: document.getElementById('btncolor').value,
        btnfont: document.getElementById('btnfontcolor').value,
    }
    apply_settings(colors);
}


js/advanced.js
function key_press(evt, div_id) {
    var id = div_id ? div_id : '';
    var c = evt ? evt.charCode : -1;
    var data;

    if (c == 61 || id == 'equals') {
        button_click('equals', 3);
        calculate();
        scroll_left();
        return true;
    }
    else if (c == 48 || id == 'zero')
        data = ['zero', '0'];
    else if (c == 49 || id == 'one')
        data = ['one', '1'];
    else if (c == 50 || id == 'two')
        data = ['two', '2'];
    else if (c == 51 || id == 'three')
        data = ['three', '3'];
    else if (c == 52 || id == 'four')
        data = ['four', '4'];
    else if (c == 53 || id == 'five')
        data = ['five', '5'];
    else if (c == 54 || id == 'six')
        data = ['six', '6'];
    else if (c == 55 || id == 'seven')
        data = ['seven', '7'];
    else if (c == 56 || id == 'eight')
        data = ['eight', '8'];
    else if (c == 57 || id == 'nine')
        data = ['nine', '9'];
    else if (c == 47 || id == 'divide')
        data = ['divide', ' / '];
    else if (c == 42 || c == 120 || id == 'multiply')
        data = ['multiply', ' * '];
    else if (c == 45 || id == 'subtract')
        data = ['subtract', ' - '];
    else if (c == 46 || id == 'dot')
        data = ['dot', '.'];
    else if (c == 43 || id == 'add')
        data = ['add', ' + '];
    else if (c == 40 || id == 'lparen')
        data = ['lparen', '('];
    else if (c == 41 || id == 'rparen')
        data = ['rparen', ')'];
    else if (c == 94 || id == 'pow')
        data = ['pow', '^'];
    else if (c == 33 || id == 'fact')
        data = ['fact', '!'];
    else if (c == 115 || id == 'sin')
        data = ['sin', 'sin('];
    else if (c == 99 || id == 'cos')
        data = ['cos', 'cos('];
    else if (c == 116 || id == 'tan')
        data = ['tan', 'tan('];
    else if (c == 83 || id == 'asin')
        data = ['asin', 'asin('];
    else if (c == 67 || id == 'acos')
        data = ['acos', 'acos('];
    else if (c == 84 || id == 'atan')
        data = ['atan', 'atan('];
    else if (c == 101 || id == 'e')
        data = ['e', 'e'];
    else if (c == 112 || id == 'pi')
        data = ['pi', 'pi'];
    else if (c == 114 || id == 'sqrt')
        data = ['sqrt', '^(1/2)'];
    else if (c == 37 || id == 'mod')
        data = ['mod', ' % '];
    else if (c == 124 || id == 'abs')
        data = ['abs', '|'];
    else if (c == 108 || id == 'log')
        data = ['log', 'log('];
    else if (c == 110 || id == 'ln')
        data = ['ln', 'ln('];
    else
        return true;

    button_click(data[0], 3);
    append_data(data[1]);
    scroll_right();
    return false;
}

function key_down(evt, div_id) {
    var id = div_id ? div_id : '';
    var k = evt ? evt.keyCode : -1;

    if (k == 27 || id == 'clr') { // escape
        button_click('clr', 1);
        if (id)
            setTimeout("button_click('clr', 2)", 150);
        reset();
    }
    else if (k == 8 || k == 46 || id == 'bksp') { // backspace/delete
        button_click('bksp', 1);
        if (id)
            setTimeout("button_click('bksp', 2)", 150);
        cut_data();
    }
    else if (k == 13) { // enter
        button_click('equals', 1);
        calculate();
        scroll_left();
    }
}

function key_up(evt) {
    var k = evt.keyCode;

    if (k == 27) // escape
        button_click('clr', 2);
    else if (k == 8 || k == 46) // backspace/delete
        button_click('bksp', 2);
    else if (k == 13) // enter
        button_click('equals', 2);
}
